File: java/src/eu/semaine/jms/SEMAINEFeatureMessage.java
package eu.semaine.jms;

import javax.jms.BytesMessage;
import javax.jms.JMSException;
import javax.jms.Message;
import javax.jms.TextMessage;

public class SEMAINEFeatureMessage extends SEMAINEMessage
{
	protected float[] features = null;
	protected String[] featureNames = null;
	
	public SEMAINEFeatureMessage(Message message)
	throws MessageFormatException
	{
		super(message);
		if (!(message instanceof TextMessage ||
			  message instanceof BytesMessage)) {
			throw new MessageFormatException("Expected either a text message or a bytes message, but got a "+message.getClass().getName());
		}
	}
	
	/**
	 * Get the feature vector from the message.
	 * @return the feature vector contained in the message.
	 * @throws JMSException
	 * @throws MessageFormatException
	 */
	public float[] getFeatureVector()
	throws JMSException, MessageFormatException
	{
		if (features == null) {
			if (message instanceof TextMessage) {
				readFromTextMessage();
			} else {
				readFromBytesMessage();
			}
		}
		assert features != null : "features are still null";
		return features;
	}

	/**
	 * Get the list of feature names if available.
	 * @return an array of strings containing the names of features if available, 
	 * or null if the feature names are not available.
	 * @throws JMSException
	 */
	public String[] getFeatureNames()
	throws JMSException
	{
		if (!(message instanceof TextMessage))
			return null; // can only get feature names from text message
		if (featureNames == null) {
			readFromTextMessage();
		}
		assert featureNames != null : "feature names are still null";
		return featureNames;
	}
	
	protected void readFromBytesMessage()
	throws JMSException
	{
		throw new RuntimeException("not yet implemented");
	}
	
	protected void readFromTextMessage()
	throws JMSException
	{
		assert message instanceof TextMessage : "this method should only be called for text messages, but message is a "+message.getClass().getName();
		String text = ((TextMessage)message).getText();
		String[] lines = text.split("\r\n|\r|\n");
		features = new float[lines.length];
		featureNames = new String[lines.length];
		for (int i=0; i<lines.length; i++) {
			// line format is:
			// value featureName
			String[] parts = lines[i].split("\\s+");
			if (parts.length > 2) {
				throw new MessageFormatException("Message contains more than two fields in line "+i+": '"+lines[i]+"'");
			}
			try {
				features[i] = Float.parseFloat(parts[0]);
			} catch (NumberFormatException nfe) {
				throw new MessageFormatException("Cannot read value '"+parts[0]+"' in line "+i+" as a float", nfe);
			}
			featureNames[i] = parts[1];
		}
	}
}

File: c++/src/semaine/cms/message/SEMAINEFeatureMessage.cpp
/*
 *  SEMAINEFeatureMessage.cpp
 *  semaine
 *
 *  Created by Marc Schröder on 17.09.08.
 *  Copyright 2008 DFKI GmbH. All rights reserved.
 *
 */

#include "SEMAINEFeatureMessage.h"

#include <typeinfo>
#include <sstream>
#include <string>

#include <cms/TextMessage.h>
#include <cms/BytesMessage.h>

namespace semaine {
namespace cms {
namespace message {

	SEMAINEFeatureMessage::SEMAINEFeatureMessage(const Message * message)
	throw(semaine::cms::exceptions::MessageFormatException) :
		SEMAINEMessage(message)
	{
		const TextMessage * tm = dynamic_cast<const TextMessage *>(message);
		const BytesMessage * bm = dynamic_cast<const BytesMessage *>(message);
		if (tm == NULL && bm == NULL) {
			throw semaine::cms::exceptions::MessageFormatException(std::string("Expected either a text message or a bytes message, but got a ")+typeid(*message).name());
		}
	}


	SEMAINEFeatureMessage::~SEMAINEFeatureMessage()
	{
	}

	std::vector<float> SEMAINEFeatureMessage::getFeatureVector()
	throw(CMSException, semaine::cms::exceptions::MessageFormatException)
	{
		if (features.empty()) {
			const TextMessage * tm = dynamic_cast<const TextMessage *>(message);
			if (tm != NULL) {
				readFromTextMessage(tm);
			} else {
				const BytesMessage * bm = dynamic_cast<const BytesMessage *>(message);
				assert (bm != NULL);
				readFromBytesMessage(bm);
			}
		}
		assert(!features.empty()); //"features are still null";
		return features;
	}

	std::vector<std::string> SEMAINEFeatureMessage::getFeatureNames()
	throw(CMSException)
	{
		if (featureNames.empty()) {
			const TextMessage * tm = dynamic_cast<const TextMessage *>(message);
			if (tm != NULL) {
				readFromTextMessage(tm);
				assert (!featureNames.empty());
			}
		}
		return featureNames;
	}
	

	void SEMAINEFeatureMessage::readFromBytesMessage(const BytesMessage * bm)
	throw(CMSException, semaine::cms::exceptions::MessageFormatException)
	{
		throw semaine::cms::exceptions::MessageFormatException("byte message not yet implemented");
	}

	

	void SEMAINEFeatureMessage::readFromTextMessage(const TextMessage * tm)
	throw(CMSException)
	{
		std::string text = tm->getText();
		std::stringstream textstream(text);
		std::string line;
		while (!getline(textstream, line).eof()) {
			std::stringstream linestream(line);
			float f;
			std::string n;
			linestream >> f >> n;
			features.push_back(f);
			featureNames.push_back(n);
		}

	}



} // namespace message
} // namespace cms
} // namespace semaine



